[WIP] Throwable Accessors

theblixguy opened the pull request:
This is a WIP PR that tracks throwable accessors: https://forums.swift.org/t/throwable-accessors/20509

Checklist:

- [x] Parsing `throws` in accessors
- [x] Type checking `try`/`try!`/`try?` statements
- [x] Type checking protocols requirements and witnesses
- [ ] Type checking edge cases
- [ ] Support for `_read` & `_modify` coroutines
- [x] Type checker tests
- [x] SIL tests

At the moment, the code compiles and works in most scenarios I have tested, but there's still work that needs to be done to ensure this feature is fully ready.

cc @rjmccall @slavapestov

Commit:
[parse] Consume throws in accessor

Commit:
[typechecker] handle throwing accessors when type checking try expressions

Commit:
[typechecker] walk subexpressions even if there's a throwing accessor

Commit:
[typechecker] look through optional expressions

Commit:
[typechecker] check if enclosing function throws or not

Commit:
[typechecker] refactor diagnosis using ASTWalker

Commit:
[typechecker] adds another walker to diagnose functions without throw

Commit:
[typechecker] move check functions inside the walker

Commit:
[codesynthesis] adds throws to synthesized accessors

Commit:
Merge branch 'master' into feature/throwable-accessors

Commit:
[astprinter] print throws if accessor throws

Commit:
[typechecker] track if the try expression is handled

Commit:
[codesynthesis] simplify throwing getter/setter code

Commit:
[codesynthesis] dont synthesize coroutine accessor body if accessor does not have storage

Commit:
[codesynthesis] fix an issue with using wrong accessor kind

Commit:
[typechecker] add checking for subscripts

Commit:
[typechecker] update diagnostic

Commit:
[test] adds a test for throwable accessors

Commit:
[test] add more test cases

Commit:
[typechecker] emit a custom diagnostic for protocol throwing accessor

Commit:
[test] update test cases

slavapestov commented:
Don’t forget to check for throws mismatches in class method overrides as well.

This looks like a good start - I’ll try to take a closer look later, and I’m sure @rjmccall will too!

Commit:
Merge branch 'master' into feature/throwable-accessors

Commit:
[typechecker] check requirement/witness read and modify accessors for throws as well

Commit:
[codesynthesis] replace comment with an explanation as to why we're returning early

Commit:
[test] add a test case for overridden accessor in class

DevAndArtist commented:
I think you should also test scenarios of protocol refinement such as the following code snippet but for accessors:

```swift
protocol P {
  func foo() throws
}

protocol Q: P {
  func foo()
}

struct S: Q {
  func foo() {
    print(42)
  }
}
```

I assume if we adapted typed throws the `Q` protocol will make specify a concrete throwing type for `foo` which in that case would be `Never` and therefore can be emitted, this also aligns nicely with accessors.

Commit:
[test] adds a protocol refinement test case

DevAndArtist commented:
I updated my example with a few more cases.

theblixguy commented:
@slavapestov Thanks, I have added a test case for a non-throwing getter/setter being overridden by a throwing one in a class - the diagnostics do need a bit of tweaking so instead of saying "method", it says "accessor".

I have also updated the protocol type checking code to check for `_read` and `_modify` as well, although I couldn't come up with a test case as I can't seem to specify the couroutines explicitly in the requirement. I tested with a class instead, however I think `TypeCheckError` needs to be updated to check for coroutines as I am currently able to override, for example, a non-throwing `_read` with a throwing one in a class.

@DevAndArtist Thanks, that's actually a good example - so this doesn't get caught:

```swift
protocol P1 {
  var foo: Int { get throws }
}

protocol Q1: P1 {
  var foo: Int { get }
}

struct S1: Q1 {
  var foo: Int {
    get { return 0 }
  }
}

let instanceS1 = S1()
let _ = (instanceS1 as P1).foo // No error
```

DevAndArtist commented:
Also take a look at the example of struct `T` above. It conforms to a protocol that requires the function to throw while the actual implementation does not throw (even the `throws` keyword can be omitted), this is similar to the refinement or overriding examples.

theblixguy commented:
@DevAndArtist Yep, that passes as well without an error diagnostic. Type checking protocols need a _bit_ more work here I guess.

Commit:
[typechecker] type check existentials properly

Commit:
[test] adds protocol comformance/refinement tests

theblixguy commented:
@DevAndArtist Fixed.

Commit:
[typechecker] type check subscripts properly too

Commit:
[test] update tests for protocol conformance/refinement

Commit:
[typechecker] handle subscripts in protocols better

Commit:
[test] add more protocol subscript tests

Commit:
[typechecker] check MemberRefExpr when used without try

Commit:
[test] add test cases for throwing accessors when used without try

Commit:
[typechecker] check for more scenarios

Commit:
[test] add more test cases

Commit:
[typechecker] check for some more non-try scenarios

Commit:
[test] add more non-try test cases

Commit:
[typechecker] remove unused variable

Commit:
[typechecker] diagnose _read coroutine overrides

Commit:
[test] adds basic coroutine tests

theblixguy commented:
I think I have covered most type checking scenarios now @slavapestov @DevAndArtist, let me know if you think these are sufficient and correct (look at the test file for examples).

Commit:
[typechecker] refactor accessor throw site checking

Commit:
Merge branch 'master' into feature/throwable-accessors

Commit:
[test] adds SILGen tests

theblixguy commented:
@slavapestov @rjmccall I have added some SILGen tests as well. What would be the next steps from here to support `_read` and `_modify` coroutines?

rjmccall commented on lib/Sema/CodeSynthesis.cpp:
```diff
@@ -340,6 +340,10 @@ createCoroutineAccessorPrototype(AbstractStorageDecl *storage,
 
+  AccessorDecl *storageAccessor = nullptr;
+  storageAccessor =
+      kind == AccessorKind::Read ? storage->getGetter() : storage->getSetter();
```
Hmm.  We should probably be following the logic we use for `mutating`-ness, which is to say that type-checking a storage declaration should compute whether its read/write/read-write operations throw and then store that in the AST, rather than trying to treat the `throws` bit on the accessor as the canonical record of this information.

theblixguy replied:
@rjmccall Thanks, I have added two bits to the storage declaration to record whether the getter/setter throws, similar to mutating-ness.

rjmccall commented on lib/Sema/CodeSynthesis.cpp:
```diff
@@ -826,6 +831,12 @@ static void synthesizeCoroutineAccessorBody(AccessorDecl *accessor,
+  // infinite loop in SILGen
+  if (!accessor->getStorage()->hasStorage()) {
+    return;
+  }
```
It's fine if something like this is necessary for incremental development, but I feel like the hack should go in SILGen, like maybe just not emitting a synthetic throwing accessor definition.

theblixguy replied:
@rjmccall Do you mean I should modify `SILGenFunction::emitCoroutineAccessor()` to not emit the couroutine accessor?

rjmccall replied:
Well, to emit it with a trivial body or something, yes.

rjmccall commented on lib/Sema/TypeCheckDeclOverride.cpp:
```diff
@@ -1672,6 +1672,20 @@ static bool checkSingleOverride(ValueDecl *override, ValueDecl *base) {
+        }
+      }
+    }
+  }
```
Why is this specific to read coroutines?

theblixguy replied:
@rjmccall The check is already handled for `_modify`, but I haven't figured out why it doesn't work for `_read` so I had to add this check.

rjmccall commented on lib/Sema/TypeCheckError.cpp:
```diff
@@ -1647,6 +1674,280 @@ class CheckErrorCoverage : public ErrorHandlingWalker<CheckErrorCoverage> {
+        return {checkComputedPropertySetter(AE, contextFlags, typeChecker,
+                                            tryLoc, isTryHandled),
+                nullptr};
+      }
```
Yeah, we really need to reintroduce `LValueAccessKind`.  That can be a separate PR.

theblixguy replied:
@rjmccall Do we need it to get this PR merged or can it be added later? I have a PR ready to bring back `LValueAccessKind` but it's been a long time since it was removed, so unfortunately the PR breaks a bunch of stuff. Maybe @slavapestov could take a look at bringing it back properly since he removed it?

rjmccall replied:
I think it's reasonable to ask Slava to help with that, yes.  But I do think we should block this PR on that one instead of adding new logic like this; there's enough of this sort of thing as it is.

theblixguy replied:
@rjmccall Okay! I read Slava's comment on SR-9960 about bringing back LValueAccessKind as some sort of reusable [AST] visitor. If Slava has some free some, then I'll let him do it, otherwise I can pick up the work if someone could point me towards how the computation is supposed to work.

theblixguy replied:
or... I can put up my PR that reverts the LValueAccessKind code that @slavapestov deleted and we can go from there - i.e. fix whatever is broken.

rjmccall commented on lib/Sema/TypeCheckProtocol.cpp:
```diff
@@ -1313,6 +1313,57 @@ bool WitnessChecker::checkWitnessAccess(ValueDecl *requirement,
+  }
+
+  return shouldDiagnose;
+}
```
This should all be based on the `doesReadThrow()` etc. bits on the storage declaration.

theblixguy replied:
@rjmccall Done!

theblixguy committed:
Merge branch 'master' into feature/throwable-accessors

theblixguy committed:
[AST] Store whether the accessors throw in the AST

theblixguy committed:
[codesynthesis] Use throws bit from storage rather than accessor

theblixguy committed:
[typechecker] use storage declaration bits to type check protocol witness accessor

theblixguy commented:
I have made some changes based on @rjmccall's feedback.

futurejones commented:
Can one of the admins verify this patch?

shahmishal commented:
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))
